### PR TITLE
[FIRRTL] Add a memory lowering pass

### PR DESCRIPTION
**Pull request by prithayan**
Create a new pass to flatten the aggregate memory instead of doing it in `LowerTypes`

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp [hunk 1]
```diff
@@ -0,0 +1,240 @@
+//===- LowerMemory.cpp - Infer Read Write Memory -----------------------===//
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp:1]: Infer read write memory

File: include/circt/Dialect/FIRRTL/Passes.td [hunk 2]
```diff
@@ -406,4 +406,10 @@ def InferReadWrite : Pass<"firrtl-infer-rw", "firrtl::FModuleOp"> {
   let constructor = "circt::firrtl::createInferReadWritePass()";
 }
 
+def LowerMemory : Pass<"firrtl-lower-memory", "firrtl::FModuleOp"> {
+  let summary = "FLatten aggregate memory data to a UInt";
```
Comment (youngar) [inline include/circt/Dialect/FIRRTL/Passes.td:410]: `FLatten`

File: lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+namespace {
+struct LowerMemoryPass : public LowerMemoryBase<LowerMemoryPass> {
+  /// This pass performs two memory transformations:
+  ///  1. If the multi-bit enable port is connected to a constant 1,
+  ///     then, replace with a single bit mask. Create a new memory with a
+  ///     1 bit mask, and replace the old memory with it. The single bit mask
+  ///     memory is always lowered to an unmasked memory.
+  ///  2. If the read and write enable ports are trivially mutually exclusive,
+  ///     then create a new memory with a single read/write port, and replace
+  ///     the old memory with it.
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp:39]: This comment might be from the InferRW pass?

File: lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAnnotations.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/Namespace.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/ADT/APSInt.h"
+#include "llvm/ADT/StringSwitch.h"
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp:21]: nit: probably APSInt and StringSwitch are not used.

File: lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+  Value getSubWhatever(ImplicitLocOpBuilder *builder, Value val, size_t index) {
+    if (BundleType bundle = val.getType().dyn_cast<BundleType>()) {
+      return builder->create<SubfieldOp>(val, index);
+    } else if (FVectorType fvector = val.getType().dyn_cast<FVectorType>()) {
+      return builder->create<SubindexOp>(val, index);
+    }
+    llvm_unreachable("Unknown aggregate type");
+    return nullptr;
+  }
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp:225]: nit: Can we move this to FIRRTLUtils and share with LowerTypes?

File: include/circt/Dialect/FIRRTL/Passes.td [hunk 6]
```diff
@@ -406,4 +406,10 @@ def InferReadWrite : Pass<"firrtl-infer-rw", "firrtl::FModuleOp"> {
   let constructor = "circt::firrtl::createInferReadWritePass()";
 }
 
+def LowerMemory : Pass<"firrtl-lower-memory", "firrtl::FModuleOp"> {
+  let summary = "Flatten aggregate memory data to a UInt";
+  let description = [{
+  }];
```
Comment (youngar) [inline include/circt/Dialect/FIRRTL/Passes.td:412]: Could use a proper description, you could probably use the comment from below:
```
  /// This pass flattens the aggregate data of memory into a UInt, and inserts
  /// appropriate bitcasts to access the data.
```

**Pull request merged**